external/audiocodec/streaming/player.c

```diff
@@ -17,627 +17,583 @@
-            bitrate = (layer == 2 /* L2 */)
-                ? kBitrateV1L2[bitrate_index - 1]
-                : kBitrateV1L3[bitrate_index - 1];
-
-            if (out_num_samples) {
-                *out_num_samples = 1152;
-            }
-        }
-		else
-		{
-            // V2 (or 2.5)
-
-            bitrate = kBitrateV2[bitrate_index - 1];
-            if (out_num_samples)
-                *out_num_samples = (layer == 1 /* L3 */) ? 576 : 1152;
-        }
-
-        if (out_bitrate)
-            *out_bitrate = bitrate;
-
-        if (version == 3 /* V1 */)
-		{
-            *frame_size = 144000 * bitrate / sampling_rate + padding;
-        }
-		else
-        {
-            // V2 or V2.5
-            size_t tmp = (layer == 1 /* L3 */) ? 72000 : 144000;
-            *frame_size = tmp * bitrate / sampling_rate + padding;
-        }
-    }
-
-    if (out_sampling_rate)
-        *out_sampling_rate = sampling_rate;
-
-    if (out_channels)
-	{
-        int channel_mode = (header >> 6) & 3;
-        *out_channels = (channel_mode == 3) ? 1 : 2;
-    }
-
-    return true;
+	RETURN_VAL_IF_FAIL((header & MP3_FRAME_VERIFY_MASK) == MP3_FRAME_VERIFY_MASK, false);
+
+	unsigned version = (header >> 19) & 3;
+	RETURN_VAL_IF_FAIL(version != 0x01, false);
+
+	unsigned layer = (header >> 17) & 3;
+	RETURN_VAL_IF_FAIL(layer != 0x00, false);
+
+	unsigned bitrate_index = (header >> 12) & 0x0f;
+	//Disallow "free" bitrate.
+	RETURN_VAL_IF_FAIL((bitrate_index != 0 && bitrate_index != 0x0f), false);
+
+	unsigned sampling_rate_index = (header >> 10) & 3;
+	RETURN_VAL_IF_FAIL((sampling_rate_index != 3), false);
+
+	static const int kSamplingRateV1[] = { 44100, 48000, 32000 };
+	int sampling_rate = kSamplingRateV1[sampling_rate_index];
+	if (version == MPEG_VERSION_2) {
+		sampling_rate /= 2;
+	} else if (version == MPEG_VERSION_2_5) {
+		sampling_rate /= 4;
+	}
+
+	unsigned padding = (header >> 9) & 1;
+
+	if (layer == MPEG_LAYER_1) {
+		static const int kBitrateV1[] = {
+			32, 64, 96, 128, 160, 192, 224, 256,
+			288, 320, 352, 384, 416, 448
+		};
+
+		static const int kBitrateV2[] = {
+			32, 48, 56, 64, 80, 96, 112, 128,
+			144, 160, 176, 192, 224, 256
+		};
+
+		int bitrate = (version == MPEG_VERSION_1)
+					  ? kBitrateV1[bitrate_index - 1]
+					  : kBitrateV2[bitrate_index - 1];
+
+		*frame_size = (12000 * bitrate / sampling_rate + padding) * 4;
+	} else {
+		static const int kBitrateV1L2[] = {
+			32, 48, 56, 64, 80, 96, 112, 128,
+			160, 192, 224, 256, 320, 384
+		};
+
+		static const int kBitrateV1L3[] = {
+			32, 40, 48, 56, 64, 80, 96, 112,
+			128, 160, 192, 224, 256, 320
+		};
+
+		static const int kBitrateV2[] = {
+			8, 16, 24, 32, 40, 48, 56, 64,
+			80, 96, 112, 128, 144, 160
+		};
+
+		int bitrate;
+		if (version == MPEG_VERSION_1) {
+			bitrate = (layer == MPEG_LAYER_2)
+					  ? kBitrateV1L2[bitrate_index - 1]
+					  : kBitrateV1L3[bitrate_index - 1];
+			*frame_size = 144000 * bitrate / sampling_rate + padding;
```

Comment:
If above 12000 and72000 & 144000 here, they are very important value please add some comment or define needed.

Reply:
To sum up, could you please do not use constant value directly?

Reply:
I fixed it.